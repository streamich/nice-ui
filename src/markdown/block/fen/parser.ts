import {
  IBoardLayout,
  IBorder,
  ILine,
  TSquarePositionComponent,
  TLineTip,
  TLineBody,
  ISquare,
} from '../../components/chess/Board/types';

const letterToNumber = {
  a: 0,
  b: 1,
  c: 2,
  d: 3,
  e: 4,
  f: 5,
  g: 6,
  h: 7,
};

const pieceName2Letter = {
  k: 'bk',
  q: 'bq',
  b: 'bb',
  n: 'bn',
  r: 'br',
  p: 'bp',
  K: 'wk',
  Q: 'wq',
  B: 'wb',
  N: 'wn',
  R: 'wr',
  P: 'wp',
};

const twoLetterToCased = {
  bk: 'k',
  bq: 'q',
  bb: 'b',
  bn: 'n',
  br: 'r',
  bp: 'p',
  wk: 'K',
  wq: 'Q',
  wb: 'B',
  wn: 'N',
  wr: 'R',
  wp: 'P',
};

const validFen = (fen: string): boolean => {
  if (typeof fen !== 'string') return false;

  // cut off any move, castling, etc info from the end
  // we're only interested in position information
  fen = fen.replace(/ .+$/, '');

  // FEN should be 8 sections separated by slashes
  const chunks = fen.split('/');
  if (chunks.length > 8) return false;

  // check the piece sections
  for (let i = 0; i < 8 && i < chunks.length; i++) {
    if (chunks[i].length > 8 || chunks[i].search(/[^kqrbnpKQRNBP1-8]/) !== -1) {
      return false;
    }
  }
  return true;
};

export const parse = (
  fen: string,
): [TSquarePositionComponent, TSquarePositionComponent, keyof typeof pieceName2Letter][] => {
  if (!validFen(fen)) return [];

  // cut off any move, castling, etc info from the end
  // we're only interested in position information
  fen = fen.replace(/ .+$/, '');

  const rows = fen.split('/');
  const pieces = [];

  for (let i = 0; i < 8 && i < rows.length; i++) {
    const row = rows[i].split('');
    let colIndex = 0;

    // loop through each character in the FEN section
    for (let j = 0; j < row.length; j++) {
      if (row[j].search(/[1-8]/) !== -1) {
        const emptySquares = parseInt(row[j], 10);
        colIndex += emptySquares;
      } else {
        let piece = row[j];
        if (piece.toLowerCase() === piece) piece = 'b' + piece.toLowerCase();
        else piece = 'w' + piece.toLowerCase();

        pieces.push([i, colIndex, piece]);
        colIndex++;
      }
    }
  }
  return pieces as any;
};

export const parseExtended = (fen: string, noFen = false): IBoardLayout => {
  const layout: IBoardLayout = {
    borders: [],
    lines: [],
    squares: [],
  };
  const parts = fen.split(/\n|;/g);

  if (!noFen) {
    const pieces = parse(parts[0]);
    if (pieces) {
      for (const [x, y, piece] of pieces) {
        layout.squares!.push({
          xy: [y, (7 - x) as any],
          piece: (twoLetterToCased as any)[piece],
        });
      }
    }
  }

  for (let i = 0; i < parts.length; i++) {
    const part = parts[i];

    // Lines like: `e1->e4`
    let match = part.match(
      /^\s*([a-hA-H])([0-8])\s*(<{0,50}|[oO]{0,1})\s*([\-=]{0,50})\s*(>{0,50}|[oO]{0,1})\s*([a-hA-H])([0-8])\s*:?\s*([a-zA-Z#0-9]+)?\s*,?\s*([0-9\.]+)?\s*/, // eslint-disable-line
    );
    if (match) {
      const line: ILine = {
        from: [(letterToNumber as any)[match[1].toLowerCase()], (Number(match[2]) - 1) as TSquarePositionComponent],
        tipFrom: match[3] as TLineTip,
        body: match[4] as TLineBody,
        tipTo: match[5] as TLineTip,
        to: [(letterToNumber as any)[match[6].toLowerCase()], (Number(match[7]) - 1) as TSquarePositionComponent],
      };
      if (match[8]) line.color = match[8];
      if (match[9]) line.opacity = Number(match[9]);
      layout.lines!.push(line);
      continue;
    }

    // Squares like: `e4:red`
    match = part.match(
      /^\s*([a-hA-H])([0-8])\s*(:(\s*([prnbkqPRNBKQ](\s|\n|$))\s*(\s*([0-9\.]+)\s*)?\s*)?\s*([\\\/\-_]{0,4})?\s*(\s*([a-zA-Z#0-9]+)?(\s*\s*([0-9\.]+)?)?)?\s*)?/, // eslint-disable-line
    );
    if (match) {
      const square: ISquare = {
        xy: [
          (Number(match[2]) - 1) as TSquarePositionComponent,
          (7 - (letterToNumber as any)[match[1].toLowerCase()]) as TSquarePositionComponent,
        ],
      };
      if (match[5]) square.piece = (pieceName2Letter as any)[match[5].trim()];
      if (match[8]) square.pieceOpacity = Number(match[8]);

      // Borders
      if (match[9]) {
        if (match[9].indexOf('-') > -1) square.borderTop = true;
        if (match[9].indexOf('\\') > -1) square.borderRight = true;
        if (match[9].indexOf('_') > -1) square.borderBottom = true;
        if (match[9].indexOf('/') > -1) square.borderLeft = true;
      }

      if (match[11]) square.color = match[11];
      if (match[13]) square.opacity = Number(match[13]);

      const existingSquare = layout.squares!.find((a) => a.xy[0] === square.xy[0] && a.xy[1] === square.xy[1]);
      if (existingSquare) {
        Object.assign(existingSquare, square);
      } else {
        layout.squares!.push(square);
      }
      continue;
    }

    // Borders like: `~a1 1 2 3 ddwwaasssd: red .7 red .2`
    match = part.match(
      /^\s*~\s*([a-hA-H])([0-8])(\s+([\-0-9\.]+))?(\s+([\-0-9\.]+))?(\s+([0-9\.]+))?(\s+([asdw]+))\s*(:\s*([a-zA-Z#0-9]+)?(\s+([0-9\.]+))?(\s+([0-9\.]+))?)?(\s+([a-zA-Z#0-9]+))?(\s+([0-9\.]+))?\s*/, // eslint-disable-line
    );
    if (match) {
      const border: IBorder = {
        xy: [(letterToNumber as any)[match[1].toLowerCase()], (Number(match[2]) - 1) as TSquarePositionComponent],
        segments: match[10],
      };
      if (match[4]) border.offset = [Number(match[4]), Number(match[6] || 0)];
      if (match[8]) border.segmentSize = Number(match[8]);
      if (match[12]) border.color = match[12];
      if (match[14]) border.opacity = Number(match[14]);
      if (match[16]) border.width = Number(match[16]);
      if (match[18]) border.fillColor = match[18];
      if (match[20]) border.fillOpacity = Number(match[20]);
      layout.borders!.push(border);
      continue;
    }

    // Size like: `size: xl`
    // match = part.match(/^\s*size\s*:\s*(xxs|xs|s|m|l|xl|xxl)(\s|\n|$)\s*/i);
    // if(match) {
    //   layout.size = match[1];
    // }
  }

  if (!layout.borders!.length && !layout.lines!.length && !layout.squares!.length) return {};
  return layout;
};
