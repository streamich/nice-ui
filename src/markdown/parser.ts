import {create} from 'md-mdast';
import {mdastToFlat} from 'mdast-flat';

const parser = create();

export const md = (markdown: string) => {
  const mdast = parser.tokenizeBlock(markdown);
  const flat = mdastToFlat(mdast as any);

  return flat;
};

export const mdi = (markdown: string) => {
  const mdast = {
    type: 'root',
    children: parser.tokenizeInline(markdown),
  };
  const flat = mdastToFlat(mdast as any);

  return flat;
};
