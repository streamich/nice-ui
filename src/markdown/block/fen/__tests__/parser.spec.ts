import {parse, parseExtended} from '../parser';

describe('parse()', () => {
  test('parses starting position', () => {
    const result = parse('rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1');
    expect(result).toEqual([
      [0, 0, 'br'],
      [0, 1, 'bn'],
      [0, 2, 'bb'],
      [0, 3, 'bq'],
      [0, 4, 'bk'],
      [0, 5, 'bb'],
      [0, 6, 'bn'],
      [0, 7, 'br'],
      [1, 0, 'bp'],
      [1, 1, 'bp'],
      [1, 2, 'bp'],
      [1, 3, 'bp'],
      [1, 4, 'bp'],
      [1, 5, 'bp'],
      [1, 6, 'bp'],
      [1, 7, 'bp'],
      [6, 0, 'wp'],
      [6, 1, 'wp'],
      [6, 2, 'wp'],
      [6, 3, 'wp'],
      [6, 4, 'wp'],
      [6, 5, 'wp'],
      [6, 6, 'wp'],
      [6, 7, 'wp'],
      [7, 0, 'wr'],
      [7, 1, 'wn'],
      [7, 2, 'wb'],
      [7, 3, 'wq'],
      [7, 4, 'wk'],
      [7, 5, 'wb'],
      [7, 6, 'wn'],
      [7, 7, 'wr'],
    ]);
  });
});

describe('parseExtended()', () => {
  test('parses starting position', () => {
    const result = parseExtended('rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1');
    expect(result).toMatchSnapshot();
  });

  test('parses lines', () => {
    const result = parseExtended('a1-->>a2\ne4 <-> g7');
    expect(result.lines).toHaveLength(2);
    expect(result.lines![0]).toEqual({
      from: [0, 0],
      tipFrom: '',
      body: '--',
      tipTo: '>>',
      to: [0, 1],
    });
    expect(result.lines![1]).toEqual({
      from: [4, 3],
      tipFrom: '<',
      body: '-',
      tipTo: '>',
      to: [6, 6],
    });
  });

  test('can add background color to existing piece square', () => {
    const result = parseExtended('K\na1:red');
    expect(result.squares![0]).toEqual({
      xy: [0, 7],
      color: 'red',
      piece: 'K',
    });
  });

  test('can add color to a square', () => {
    const result = parseExtended('b2:red');
    expect(result.squares![0]).toEqual({
      xy: [1, 6],
      color: 'red',
    });
  });
});
