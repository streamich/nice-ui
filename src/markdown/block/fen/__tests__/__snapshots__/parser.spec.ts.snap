// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseExtended() parses starting position 1`] = `
{
  "borders": [],
  "lines": [],
  "squares": [
    {
      "piece": "r",
      "xy": [
        0,
        7,
      ],
    },
    {
      "piece": "n",
      "xy": [
        1,
        7,
      ],
    },
    {
      "piece": "b",
      "xy": [
        2,
        7,
      ],
    },
    {
      "piece": "q",
      "xy": [
        3,
        7,
      ],
    },
    {
      "piece": "k",
      "xy": [
        4,
        7,
      ],
    },
    {
      "piece": "b",
      "xy": [
        5,
        7,
      ],
    },
    {
      "piece": "n",
      "xy": [
        6,
        7,
      ],
    },
    {
      "piece": "r",
      "xy": [
        7,
        7,
      ],
    },
    {
      "piece": "p",
      "xy": [
        0,
        6,
      ],
    },
    {
      "piece": "p",
      "xy": [
        1,
        6,
      ],
    },
    {
      "piece": "p",
      "xy": [
        2,
        6,
      ],
    },
    {
      "piece": "p",
      "xy": [
        3,
        6,
      ],
    },
    {
      "piece": "p",
      "xy": [
        4,
        6,
      ],
    },
    {
      "piece": "p",
      "xy": [
        5,
        6,
      ],
    },
    {
      "piece": "p",
      "xy": [
        6,
        6,
      ],
    },
    {
      "piece": "p",
      "xy": [
        7,
        6,
      ],
    },
    {
      "piece": "P",
      "xy": [
        0,
        1,
      ],
    },
    {
      "piece": "P",
      "xy": [
        1,
        1,
      ],
    },
    {
      "piece": "P",
      "xy": [
        2,
        1,
      ],
    },
    {
      "piece": "P",
      "xy": [
        3,
        1,
      ],
    },
    {
      "piece": "P",
      "xy": [
        4,
        1,
      ],
    },
    {
      "piece": "P",
      "xy": [
        5,
        1,
      ],
    },
    {
      "piece": "P",
      "xy": [
        6,
        1,
      ],
    },
    {
      "piece": "P",
      "xy": [
        7,
        1,
      ],
    },
    {
      "piece": "R",
      "xy": [
        0,
        0,
      ],
    },
    {
      "piece": "N",
      "xy": [
        1,
        0,
      ],
    },
    {
      "piece": "B",
      "xy": [
        2,
        0,
      ],
    },
    {
      "piece": "Q",
      "xy": [
        3,
        0,
      ],
    },
    {
      "piece": "K",
      "xy": [
        4,
        0,
      ],
    },
    {
      "piece": "B",
      "xy": [
        5,
        0,
      ],
    },
    {
      "piece": "N",
      "xy": [
        6,
        0,
      ],
    },
    {
      "piece": "R",
      "xy": [
        7,
        0,
      ],
    },
  ],
}
`;
